test/unit/DominantJuice.unit.t.sol

```diff
@@ -0,0 +1,962 @@
+        // vm.mockCall(
+        //     address(directory),
+        //     abi.encodeCall(directory.isTerminalOf, (projectID, ethPaymentTerminal)),
+        //     abi.encode(true)
+        // );
+        vm.warp(CYCLE_DURATION + 60);
+        vm.expectRevert(DominantJuice.CycleHasExpired.selector);
+        vm.prank(address(ethPaymentTerminal));
+        dominantJuice.payParams(payParamsData);
+    }
+
+    // PASSING
+    function testFuzz_payParams_revertsWhenPaymentIsBelowMinPledge(uint256 _value) public bonusDeposited {
+        // In reality, payParams() will be called by the JB Payment Terminal Store, but mocking a direct call is
+        // a good check for unit testing purposes.
+        vm.assume(_value < MIN_PLEDGE_AMOUNT);
+        payParamsData.amount.value = _value;
+
+        vm.expectRevert(abi.encodeWithSelector(DominantJuice.AmountIsBelowMinimumPledge.selector, MIN_PLEDGE_AMOUNT));
+        vm.prank(rando);
+        dominantJuice.payParams(payParamsData);
+    }
+
+    // PASSING - (unimplemented feature error)
+    function test_payParams_happyPathReturnsMemoryVariables() public bonusDeposited {
+        payParamsData.payer = pledger1;
+        payParamsData.weight = 1;
+        payParamsData.memo = "juice";
+        payParamsData.amount = JBTokenAmount(ethToken, MIN_PLEDGE_AMOUNT, 18, 1);
+
+        // Error: Unimplemented feature (/solidity/libsolidity/codegen/ArrayUtils.cpp:228):
+        // Copying of type struct JBPayDelegateAllocation3_1_1 memory[] memory to storage not yet supported.
+        // delegateAllocations = new JBPayDelegateAllocation3_1_1[](1);
+        // delegateAllocations[0] = JBPayDelegateAllocation3_1_1(dominantJuice, 0, "");
+
+        (uint256 _weight, string memory _memo,) = dominantJuice.payParams(payParamsData);
+        assertEq(payParamsData.weight, _weight);
+        assertEq(payParamsData.memo, _memo);
+        //assertEq(0, delegateAllocations[0].amount);
+
+        // Assert that there are no major contract changes when payParams is called directly:
+        DominantJuice.FundingStatus memory status = dominantJuice.getCycleFundingStatus();
+
+        assertEq(0, status.totalPledged);
+        assertEq(0, status.percentOfGoal);
+        assertFalse(status.isTargetMet);
+        assertFalse(status.hasCycleExpired);
+        assertEq(TOTAL_REFUND_BONUS, dominantJuice.getBalance());
+        (uint256 pledgerW, uint256 totalW) = dominantJuice.getPledgerAndTotalWeights(pledger1);
+        assertEq(0, pledgerW);
+        assertEq(0, totalW);
+    }
+
+    // payParams() function is non-payable, the parent function that calls it is nonReentrant, and there are only if
+    // statements and logic to satisfy Juicebox architecture, thus not many unit tests here. This function can be
+    // called directly, but nothing would happen. The only way to pledge correctly is by calling
+    // `JBPayoutRedemptionPaymentTerminal3_1_1.pay()`.
+
+    ///////////////////////
+    // didPay() Tests
```

Comment:
So you do think I should add these if statements then to `didPay()`?: 
```if (getBalance() == 0) revert RefundBonusNotDeposited();
 if (hasCycleExpired()) revert CycleHasExpired();```
 
 And this would be in case JB architecture is functioning properly right? I considered it before as well, and I think it’s good to have as an edge case backup, and I would put them in a modifier then since they would be on both functions. 
 
Here's a good specific question that may help me understand more with where the line is of too shallow versus too deep tests: For the `payParams()` minimum pledge amount check, would you add to `didPay()` as well? The pledge has already been made at this point, and `payParams()` is “in front” of `didPay()` in the chain (and didPay can only be called by a Terminal). The only way it would not be the same is a major glitch mid-call chain in JB architecture. So yeah, how deep of an error do we need to consider could happen? What is too deep?

Reply:
I think I was too on-the-fence in the google doc when we were talking about it, and also I hadn't fully grokked this code yet. Since it's a contract, on-chain should be where all validations are done. It's much more important that these are done in mutating functions (vs view), so if you're going to be putting logic like this in view functions, they definitely should be in the mutating functions as well. So, yeah, I guess the answer is that these checks should be done in didPay